src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+		  sprintf(gr->TK0.init, "%s", TK0init);
+		  gr->TK1.align = TK1->align;
+		  gr->TK1.size  = TK1->size;
+		  sprintf(gr->TK1.defn, "%s", TK1->cluda_name);
+		  sprintf(gr->TK1.init, "0");
+		break;
+		case GA_REDUCE_MAXANDARGMAX:
+		case GA_REDUCE_ARGMAX:
+		case GA_REDUCE_MAX:
+		  TK0 = TPS0;
+		  TK1 = gpuarray_get_type(gr->TI0tc);
+		  reduxGetMaxInit (TK0->typecode, &TK0init);
+		  gr->TK0.align = TK0->align;
+		  gr->TK0.size  = TK0->size;
+		  sprintf(gr->TK0.defn, "%s", TK0->cluda_name);
+		  sprintf(gr->TK0.init, "%s", TK0init);
+		  gr->TK1.align = TK1->align;
+		  gr->TK1.size  = TK1->size;
+		  sprintf(gr->TK1.defn, "%s", TK1->cluda_name);
+		  sprintf(gr->TK1.init, "0");
+		break;
+		case GA_REDUCE_ALL:
+		case GA_REDUCE_AND:
+		  TK0 = TPS0;
+		  reduxGetAndInit (TK0->typecode, &TK0init);
+		  gr->TK0.align = TK0->align;
+		  gr->TK0.size  = TK0->size;
+		  sprintf(gr->TK0.defn, "%s", TK0->cluda_name);
+		  sprintf(gr->TK0.init, "%s", TK0init);
+		break;
+		case GA_REDUCE_ANY:
+		case GA_REDUCE_XOR:
+		case GA_REDUCE_OR:
+		  TK0 = TPS0;
+		  reduxGetOrInit  (TK0->typecode, &TK0init);
+		  gr->TK0.align = TK0->align;
+		  gr->TK0.size  = TK0->size;
+		  sprintf(gr->TK0.defn, "%s", TK0->cluda_name);
+		  sprintf(gr->TK0.init, "%s", TK0init);
+		break;
+		default:
+		  ;/* Unreachable */
+	}
+}
+
+/**
+ * Iterate over the arguments of the reduction operator.
+ */
+
+static void        reduxGenIterArgs                 (const GpuReduction*  gr,
+                                                     GpuReductionIterFn   fn,
+                                                     void*                user){
+	int k;
+	
+	/**
+	 * Template selector
+	 */
+	
+	fn(gr, gr->TU32tc, "TU32",                              "selector",    0, user);
+	
+	/**
+	 * "Universal" parameters describing the partitioning of the problem.
+	 */
+	
+	fn(gr, gr->TU64tc, "TU64",                              "U",           0, user);
+	fn(gr, gr->TU64tc, "TU64",                              "V",           0, user);
+	fn(gr, gr->TU64tc, "TU64",                              "B",           0, user);
+	fn(gr, gr->TU32tc, "TU32",                              "D",           0, user);
+	fn(gr, gr->TU32tc, "TU32",                              "Dunit",       0, user);
+	fn(gr, gr->TU32tc, "TU32",                              "H",           0, user);
+	
+	/* Global Lattice Coordinates */
+	fn(gr, gr->TU32tc, "TU32",                              "LSlice",      0, user);
+	fn(gr, gr->TU32tc, "TU64",                              "LPadded",     0, user);
+	for (k=0;k < gr->nds;k++){
+		fn(gr, gr->TU64tc, "TU64",                              "L%d",         k, user);
+	}
+	for (k=0;k < gr->log2MaxBS;k++){
+		fn(gr, gr->TU32tc, "TU32",                              "L%di",        k, user);
+	}
+	
+	/* S0 Lattice */
+	if (reduxGenKernelRequiresLatticeS0(gr)){
+		fn(gr, GA_BUFFER,  "const GLOBAL_MEM char* restrict",   "S0",          0, user);
```

Comment:
OpenCL 1.0, _§6.1.4 Reserved Data Types_: _"The type qualifiers `const`, `restrict` and `volatile` as defined by the C99 specification are supported. These qualifiers cannot be used with `image2d_t` and `image3d_t` type. Types other than pointer types shall not use the `restrict` qualifier."_